src/tokens/stream_reader.rs

```diff
@@ -0,0 +1,539 @@
+where
+    T: TokenStream<'a>,
+{
+    stream: T,
+    /// Models what containers we are in
+    container_stack: Vec<ContainerType>,
+
+    // XXX this is a RefCell<AnnotationToken> because we need interior mutability for memoization
+    /// the current token
+    curr_token_cell: Option<RefCell<Token<'a>>>,
+
+    /// remember the current read item
+    curr_item: StreamItem,
+
+    /// memory set aside to support `read_str` since we cannot safely return a reference into the
+    /// `RefCell` that holds the token without wrapping it in a `Ref`.
+    str_buf: Str,
+}
+
+impl<'a, T> TokenStreamReader<'a, T>
+where
+    T: TokenStream<'a>,
+{
+    /// Advances the stream, setting the current token and item.
+    fn next_token(&mut self, instruction: Instruction) -> IonResult<StreamItem> {
+        let token = self.stream.next_token(instruction)?;
+        let item = match &token.content {
+            Content::Null(ion_type) => StreamItem::Null(*ion_type),
+            Content::Scalar(thunk) => StreamItem::Value(thunk.scalar_type().into()),
+            Content::StartContainer(container_type) => StreamItem::Value((*container_type).into()),
+            Content::EndContainer(_) => StreamItem::Nothing,
+            Content::EndStream => StreamItem::Nothing,
+        };
+        self.curr_item = item;
+        self.curr_token_cell = Some(RefCell::new(token));
+        Ok(item)
+    }
+
+    /// Resets the state to just before something at the current level of the stream
+    fn reset_curr_state(&mut self) -> IonResult<()> {
+        self.curr_item = StreamItem::Nothing;
+        self.curr_token_cell = None;
+        Ok(())
+    }
+}
+
+impl<'a, T> From<T> for TokenStreamReader<'a, T>
+where
+    T: TokenStream<'a>,
+{
+    fn from(stream: T) -> Self {
+        TokenStreamReader {
+            stream,
+            container_stack: vec![],
+            curr_token_cell: None,
+            curr_item: StreamItem::Nothing,
+            str_buf: Str::from(""),
+        }
+    }
+}
+
+const STEP_IN_ERROR_TEXT: &str = "Cannot step in, not at start of container";
+const STEP_OUT_ERROR_TEXT: &str = "Cannot step out, not in a container";
+const NO_FIELD_NAME_ERROR_TEXT: &str = "No field name";
+const NOT_POSITIONED_ON_ANYTHING_ERROR_TEXT: &str = "Not positioned on anything";
+const CANNOT_READ_NON_SCALAR_ERROR_TEXT: &str = "Cannot read from non-scalar";
+
+/// Generates the read methods against the underlying token/item state.
+/// This eliminates the boilerplate around each definition avoiding a lot of copy/paste.
+///
+/// * `me` - the identifier for `self` to make it visible to `scalar_exp`.
+/// * `method` - the name of the method.
+/// * `scalar_type` - the return type for the method.
+/// * `variant` - the unqualified variant of [`ScalarValue`] to match over the current token.
+/// * `scalar` - the matched identifier of the contents of [`ScalarValue`].
+/// * `scalar_exp` - the expression to generate the return value.
+macro_rules! read_method_self {
```

Comment:
🗺️ It makes me sad to do this, but this macro really does avoid a lot of boilerplate around implementing the `read_XXX` methods especially given our `Option<RefCell<Token>>` modalities.